mypy/solve.py

```diff
@@ -33,62 +41,332 @@ def solve_constraints(
-        elif bottom is None:
-            if top:
-                candidate = top
+        if top is None:
+            top = target
+        else:
+            top = meet_types(top, target)
+
+    p_top = get_proper_type(top)
+    p_bottom = get_proper_type(bottom)
+    if isinstance(p_top, AnyType) or isinstance(p_bottom, AnyType):
+        source_any = top if isinstance(p_top, AnyType) else bottom
+        assert isinstance(source_any, ProperType) and isinstance(source_any, AnyType)
+        return AnyType(TypeOfAny.from_another_any, source_any=source_any)
+    elif bottom is None:
+        if top:
+            candidate = top
+        else:
+            # No constraints for type variable
+            return None
+    elif top is None:
+        candidate = bottom
+    elif is_subtype(bottom, top):
+        candidate = bottom
+    else:
+        candidate = None
+    return candidate
+
+
+def normalize_constraints(
+    constraints: list[Constraint], vars: list[TypeVarId]
+) -> list[Constraint]:
+    """Normalize list of constraints (to simplify life for the non-linear solver).
+
+    This includes two things currently:
+      * Complement T :> S by S <: T
+      * Remove strict duplicates
+    """
+    res = constraints.copy()
+    for c in constraints:
+        if isinstance(c.target, TypeVarType):
+            res.append(Constraint(c.target, neg_op(c.op), c.origin_type_var))
+    return [c for c in remove_dups(constraints) if c.type_var in vars]
+
+
+def propagate_constraints_for(
+    var: TypeVarId, direction: int, cmap: dict[TypeVarId, list[Constraint]]
+) -> list[Constraint]:
+    """Propagate via linear constraints to get additional constraints for `var`.
+
+    For example if we have constraints:
+        [T <: int, S <: T, S :> str]
+    we can add two more
+        [S <: int, T :> str]
+    """
+    extra_constraints = []
+    seen = set()
+    front = [var]
+    if cmap[var]:
+        var_def = cmap[var][0].origin_type_var
+    else:
+        return []
+    while front:
+        tv = front.pop(0)
+        for c in cmap[tv]:
+            if (
+                isinstance(c.target, TypeVarType)
+                and c.target.id not in seen
+                and c.target.id in cmap
+                and c.op == direction
+            ):
+                front.append(c.target.id)
+                seen.add(c.target.id)
+            elif c.op == direction:
+                new_c = Constraint(var_def, direction, c.target)
+                if new_c not in cmap[var]:
+                    extra_constraints.append(new_c)
+    return extra_constraints
+
+
+def transitive_closure(
+    tvars: list[TypeVarId], constraints: list[Constraint]
+) -> tuple[dict[TypeVarId, set[Type]], dict[TypeVarId, set[Type]]]:
+    """Find transitive closure for given constraints on type variables.
+
+    Transitive closure gives maximal set of lower/upper bounds for each type variable, such
+    we cannot deduce any further bounds by chaining other existing bounds.
```

Comment:
Give an example or two that illustrates what this means in the docstring? Again, it would be good to have some unit tests to validate the logic and avoid regressions in the future.